presto-spi/src/main/java/com/facebook/presto/spi/plan/CteReferenceNode.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.plan;
+
+import com.facebook.presto.spi.SourceLocation;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Collections.singletonList;
+import static java.util.Objects.requireNonNull;
+
+@Immutable
+public final class CteReferenceNode
+        extends PlanNode
+{
+    private final PlanNode source;
+    private final String cteName;
+
+    @JsonCreator
+    public CteReferenceNode(
+            Optional<SourceLocation> sourceLocation,
+            @JsonProperty("id") PlanNodeId id,
+            @JsonProperty("source") PlanNode source,
+            @JsonProperty("cteName") String cteName)
+    {
+        this(sourceLocation, id, Optional.empty(), source, cteName);
+    }
+
+    public CteReferenceNode(
+            Optional<SourceLocation> sourceLocation,
+            PlanNodeId id,
+            Optional<PlanNode> statsEquivalentPlanNode,
+            PlanNode source,
+            String cteName)
+    {
+        super(sourceLocation, id, statsEquivalentPlanNode);
+        this.cteName = requireNonNull(cteName, "cteName must not be null");
+        this.source = requireNonNull(source, "source must not be null");
+    }
+
+    @Override
+    public List<PlanNode> getSources()
+    {
+        return singletonList(source);
+    }
+
+    @JsonProperty
+    public PlanNode getSource()
+    {
+        return source;
+    }
+
+    @Override
+    public List<VariableReferenceExpression> getOutputVariables()
+    {
+        return source.getOutputVariables();
+    }
+
+    @Override
+    public PlanNode replaceChildren(List<PlanNode> newChildren)
+    {
+        requireNonNull(newChildren);
```

Comment:
Nit: please add the message too, so it's very obvious this is a validation failure and therefore very indicative of something that wasn't considered during development.
```suggestion
        requireNonNull(newChildren, "newChildren is null");
```

Please make sure the message is consistent, we use one convention which is `VARIABLE_NAME is null`.